packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/CustomTabsConnectionHelper.java

```diff
@@ -0,0 +1,115 @@
+package expo.modules.webbrowser;
+
+import android.content.ComponentName;
+import android.content.Context;
+import android.net.Uri;
+import android.support.customtabs.CustomTabsClient;
+import android.support.customtabs.CustomTabsServiceConnection;
+import android.support.customtabs.CustomTabsSession;
+
+import expo.core.interfaces.LifecycleEventListener;
+
+public class CustomTabsConnectionHelper extends CustomTabsServiceConnection implements LifecycleEventListener {
+
+  private Context mContext;
+  private String mPackageName;
+  private DeferredClientActionsQueue<CustomTabsClient> clientActions = new DeferredClientActionsQueue<>();
+  private DeferredClientActionsQueue<CustomTabsSession> sessionActions = new DeferredClientActionsQueue<>();
+
+  CustomTabsConnectionHelper(Context context) {
+    this.mContext = context;
+  }
+
+  void warmUp(String packageName) {
+    clientActions.executeOrQueueAction(client -> client.warmup(0));
+    ensureConnection(packageName);
+  }
+
+  void mayInitWithUrl(String packageName, Uri uri) {
+    sessionActions.executeOrQueueAction(session -> session.mayLaunchUrl(uri, null, null));
+    ensureConnection(packageName);
+    ensureSession();
+  }
+
+  private void ensureSession() {
+    if (!sessionActions.hasClient()) {
+      clientActions.executeOrQueueAction(
+          client ->
+              sessionActions.setClient(client.newSession(null)));
+    }
+  }
+
+  boolean coolDown(String packageName) {
+    if (packageName.equals(this.mPackageName)) {
+      unbindService();
+      return true;
+    }
+    return false;
+  }
+
+  private void ensureConnection(String packageName) {
+    if (this.mPackageName != null && !this.mPackageName.equals(packageName)) {
+      clearConnection();
+    }
+    if (!connectionStarted(packageName)) {
+      CustomTabsClient.bindCustomTabsService(getContext(), packageName, this);
+      this.mPackageName = packageName;
+    }
+  }
+
+  private boolean connectionStarted(String packageName) {
+    return packageName.equals(this.mPackageName);
+  }
+
+  @Override
+  public void onBindingDied(ComponentName componentName) {
+    if (componentName.getPackageName().equals(this.mPackageName)) {
+      clearConnection();
+    }
+  }
+
+  @Override
+  public void onCustomTabsServiceConnected(ComponentName componentName, CustomTabsClient client) {
```

Comment:
Would it make sense to instead of enqueueing — resolve the promise only after the service connects? 🤔

Reply:
We would still need enqueuing though.

The reason I haven't done this is the fact, that it might be unreliable, based on my experience. Hence, it might happen that we never get connected.

And it makes sense in some way. Truth is, that those methods are not designed to be 100% reliable, just to speedUp the most likely scenario. I am simply afraid that for some browser we might end up with never resolved promise.